design/metadata-handling.md

```diff
@@ -0,0 +1,417 @@
+
+- **machineName** : returns the Machine name
+- **metal3MachineName** : returns the Metal3Machine name
+- **bareMetalHostName** : returns the BareMetalHost name
+- **index** : returns the Metal3Machine index for the Metal3Metadata object.
+  The index starts from 0.
+- **indexWithOffset** : takes an integer as parameter and returns the sum of
+  the index and the offset parameter
+- **indexWithStep** : takes an integer as parameter and returns the
+  multiplication of the index and the step parameter
+- **indexWithOffsetAndStep** OR **indexWithStepAndOffset**: takes two
+  integers as parameters, order depending on the function name, and returns the
+  sum of the offset and the multiplication of the index and the step.
+- **index*Hex** : All the `index` functions can be suffixed with `Hex` to
+  get the same value in hexadecimal format.
+- **bareMetalHostMACByName**: takes a string as parameter and returns the MAC
+  address of the nic with the name matching the parameter. This function
+  operates over the list of NICs in the `status.hardwareDetails.NIC` field of
+  the BareMetalHost.
+
+
+The output of the controller would be secrets, one per node linking to the
+Metal3DataTemplate object.
+
+If the Metal3DataTemplate object is updated, the generated secrets will not be
+updated, to allow for reprovisioning of the nodes in the exact same state as
+they were initially provisioned. Hence, to do an update, it is necessary to do
+a rolling upgrade of all nodes.
+
+The reconciliation of the Metal3DataTemplate object will also be triggered by
+changes on Metal3Machines. In the case that a Metal3Machine gets modified, if
+the `dataTemplate` references a Metal3Metadata, that object will be reconciled.
+If the `metaData` and `networkData` are set, that will be a no-op reconciliation
+loop. If it is unset, there will be two cases:
+
+- An already generated secret exists with an ownerReference to this
+  Metal3Machine. In that case, the reconciler will update it and fill the
+  respective field on the Metal3Machine with the secret name.
+- if no secret exists with an ownerReference to this Metal3Machine, then the
+  reconciler will create one and fill the respective field with the secret name.
+
+To create a secret, the controller will generate the content
+based on the `metaData` or `networkData` field of the Metal3DataTemplate Specs.
+It will render the content of the secret, selecting the lowest available index
+from the status indexes map. If that map is empty, it will fill it by building
+a map of names of the machines and index used by the machine, extracting the
+index from the secret names in the Metal3Machine specs.
+
+Once the next available index is found, it will update the Metal3DataTemplate
+object. Upon conflict, it will immediately requeue to consider the new state of
+Metal3DataTemplate. Upon success, it will render the content values, and create
+a secret containing the rendered data.
```

Comment:
I'm still unclear why this is so focused on indexes. You say random values aren't sufficient, but don't explain how. Could you provide more detail about that, please?